active/0021-msg-batch-for-resources.md

```diff
@@ -0,0 +1,152 @@
+
+![Old Data Plane](0021-assets/resource-old-arch-data-plane.png)
+
+When a message or an event is triggered in emqx, the MQTT connection process (in the PubSub layer) calls back the components in the second layer through `emqx_hooks`.
+
+The second layer is the components related to external resources, such as Data Bridge, Authentication (AuthN), Authorization (AuthZ), etc. They calls `emqx_resource:query/3` to query the resources.
+
+The `emqx_resource:query/3` directly calls back the DB drivers in the bottom layer, no messaging passing is needed.
+
+#### Managing Resources through the Resource Layer
+
+On the other hand, when the user creates resources via HTTP APIs, it calls creation APIs of the second layer (Data Bridges, AuthN and AuthZ), and then they call `emqx_resource:create/3` in the third layer.
+
+The `emqx_resource:create/3` creates a resource manager process, all the subsequent management operations will go through the manager process:
+
+![Old Control Plane](0021-assets/resource-old-arch-control-plane.png)
+
+The resource management calls through all the layers from the top to the bottom.
+
+### The Layers After the Change
+
+We suggest introduce the "resource workers" into the resource layer.
+So that the old resource layer is divided into two parts: "resource workers" and "resource management". 
+
+The "resource workers" is a pool maintains several worker processes.
+It is the component for querying external resources, maintaining the message queue (`replayq`), and sending the messages to the drivers. 
+
+The "resource management" part is the old resource layer, and remains unchanged as before, which is responsible for resource management operations.
+
+#### Sending Messages through the Resource Workers
+
+![New Data Plane](0021-assets/resource-new-arch-data-plane.png)
+
+#### Managing Resources through the Resource Manager
+
+![New Data Plane](0021-assets/resource-new-arch-control-plane.png)
+
+### Pool of Resource Workers with ReplayQs
+
+In the current implementation, each time a resource is created, a resource manager process will be created for each resource ID, which is responsible for maintaining the relevant state of the resource. See the code of `emqx_resource_manager` module for details.
+
+After the implementation of message queue is added, we also create a resource worker pool each time a resource is created, which is responsible for the process of accessing external resources and message sending.
+
+The following figure is a schematic diagram of the resource worker pool. Each worker maintains a ReplayQ:
+
+![Resource Workers with ReplayQs](0021-assets/resource-workers.png)
+
+The messages are first saved by the worker to the queue (which can be memory or disk queue), and then according to the batching policy, the worker takes the message out of the queue and sends it to the corresponding driver through the connector callback modules.
```

Comment:
Speak of using QUEUE. we could have producer and consumer.

Resource worker as producer push data to the QUEUE.
Another process Driver worker or the batcher periodically pop(or just read) data from the queue from head so that Producer and Consumer they will not interfere each other.

Also we should ensure the data poped could only disappear after the data has been written to the db backend so we will not loss data in case backend worker crash or restart?

Reply:
Good suggestion, It is good to have producer workers and consumer workers. I will update the doc for this.

For the `pop`, if I understand correctly, the `replayq:pop()` won't delete the items from disk until they get ACKed using `replayq:ack()`. I will check this later, if it doesn't work as expected, I will first `peek` and then `pop`.